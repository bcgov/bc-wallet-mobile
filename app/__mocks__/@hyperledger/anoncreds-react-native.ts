const AnonCredsReactNative = {
  install: jest.fn(),
}

export default AnonCredsReactNative
