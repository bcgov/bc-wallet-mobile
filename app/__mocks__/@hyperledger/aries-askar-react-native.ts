const AriesAskarReactNative = {
  install: jest.fn(),
}

export default AriesAskarReactNative
