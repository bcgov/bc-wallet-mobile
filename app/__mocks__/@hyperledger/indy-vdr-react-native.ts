const IndyVDRReactNative = {
  install: jest.fn(),
}

export default IndyVDRReactNative
