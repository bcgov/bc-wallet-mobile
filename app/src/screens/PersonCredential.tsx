import { useAgent } from '@aries-framework/react-hooks'
import { useStore, useTheme, Button, ButtonType, testIdWithKey } from '@hyperledger/aries-bifold-core'
import React, { useState, useCallback, useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View, TouchableOpacity, Linking, Platform, ScrollView } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import Icon from 'react-native-vector-icons/MaterialIcons'

import PersonIssuance1 from '../assets/img/PersonIssuance1.svg'
import PersonIssuance2 from '../assets/img/PersonIssuance2.svg'
import LoadingIcon from '../components/LoadingIcon'
import { startFlow } from '../helpers/BCIDHelper'
import { useCredentialOfferTrigger } from '../hooks/credential-offer-trigger'
import { BCState } from '../store'

export default function PersonCredential() {
  const { agent } = useAgent()

  const [store] = useStore<BCState>()
  const [appInstalled, setAppInstalled] = useState<boolean>(false)
  const [workflowInProgress, setWorkflowInProgress] = useState<boolean>(false)
  const [workflowConnectionId, setWorkflowConnectionId] = useState<string | undefined>()

  const { ColorPallet, TextTheme } = useTheme()
  const { t } = useTranslation()

  useCredentialOfferTrigger(workflowConnectionId)

  const styles = StyleSheet.create({
    pageContainer: {
      flex: 1,
    },
    pageContent: {
      flex: 1,
      flexGrow: 1,
    },
    credentialCardContainer: {
      marginVertical: 20,
      display: 'flex',
      alignItems: 'center',
    },
    button: {
      marginBottom: 15,
    },
    section: {
      backgroundColor: ColorPallet.brand.secondaryBackground,
      paddingVertical: 24,
      paddingHorizontal: 25,
      marginTop: 10,
      flexGrow: 1,
    },
    sectionHeader: {
      flexDirection: 'row',
      alignItems: 'center',
      paddingBottom: 10,
      display: 'flex',
    },
    sectionSecondaryAction: {
      display: 'flex',
      alignItems: 'center',
      marginTop: 10,
    },
    sectionSeparator: {
      marginBottom: 10,
    },
    sectionRow: {
      flexDirection: 'row',
      alignItems: 'center',
      flexGrow: 1,
      paddingHorizontal: 25,
    },
    itemSeparator: {
      borderBottomWidth: 1,
      borderBottomColor: ColorPallet.brand.primaryBackground,
      marginHorizontal: 25,
    },
  })

  const isBCServicesCardInstalled = async () => {
    return await Linking.canOpenURL('ca.bc.gov.id.servicescard://')
  }

  useEffect(() => {
    isBCServicesCardInstalled().then((result) => {
      setAppInstalled(result)
    })
  }, [])

  const acceptPersonCredentialOffer = useCallback(() => {
    setWorkflowInProgress(true)
    startFlow(agent!, store, setWorkflowInProgress, t, (connectionId) => setWorkflowConnectionId(connectionId))
  }, [])

  const getBCServicesCardApp = useCallback(() => {
    setAppInstalled(true)
    const url =
      Platform.OS === 'ios'
        ? 'https://apps.apple.com/us/app/id1234298467'
        : 'https://play.google.com/store/apps/details?id=ca.bc.gov.id.servicescard'
    return Linking.openURL(url)
  }, [])

  return (
    <SafeAreaView style={styles.pageContainer} edges={['bottom', 'left', 'right']}>
      <ScrollView style={styles.pageContent}>
        <View style={styles.credentialCardContainer}>{appInstalled ? <PersonIssuance2 /> : <PersonIssuance1 />}</View>
        <View style={styles.section}>
          <View style={styles.sectionHeader}>
            <Text
              accessibilityRole={'header'}
              style={[
                TextTheme.headingThree,
                {
                  flexShrink: 1,
                  color: appInstalled ? ColorPallet.brand.primaryDisabled : TextTheme.headingThree.color,
                },
              ]}
            >
              {appInstalled ? t('PersonCredential.ServicesCardInstalled') : t('PersonCredential.InstallServicesCard')}
            </Text>
            {appInstalled && (
              <Icon
                name="check-circle"
                testID={testIdWithKey('AppInstalledIcon')}
                size={35}
                style={{ marginLeft: 10, color: ColorPallet.semantic.success }}
              />
            )}
          </View>
          {appInstalled ? null : (
            <View style={{ marginTop: 10 }}>
              <Button
                buttonType={ButtonType.Primary}
                onPress={getBCServicesCardApp}
                accessibilityLabel={t('PersonCredential.InstallApp')}
                testID={testIdWithKey('InstallApp')}
                title={t('PersonCredential.InstallApp')}
              ></Button>
              <TouchableOpacity
                onPress={() => setAppInstalled(true)}
                accessibilityLabel={t('PersonCredential.AppOnOtherDevice')}
                testID={testIdWithKey('AppOnOtherDevice')}
                style={styles.sectionSecondaryAction}
              >
                <Text style={{ ...TextTheme.bold, color: ColorPallet.brand.primary }}>
                  {t('PersonCredential.AppOnOtherDevice')}
                </Text>
              </TouchableOpacity>
            </View>
          )}
        </View>
        <View style={styles.section}>
          <View style={styles.sectionHeader}>
            <Text
              accessibilityRole={'header'}
              style={[
                TextTheme.headingThree,
                {
                  flexShrink: 1,
                  color: appInstalled ? TextTheme.headingThree.color : ColorPallet.brand.primaryDisabled,
                },
              ]}
            >
              {t('PersonCredential.CreatePersonCred')}
            </Text>
          </View>
          {appInstalled ? (
            <Button
              buttonType={ButtonType.Primary}
              testID={testIdWithKey('StartProcess')}
              accessibilityLabel={t('PersonCredential.StartProcess')}
              title={t('PersonCredential.StartProcess')}
              onPress={acceptPersonCredentialOffer}
            >
              <View style={{ opacity: workflowInProgress ? 1 : 0, marginLeft: -18 }}>
                <LoadingIcon active={workflowInProgress} size={35} color={ColorPallet.brand.buttonText} />
              </View>
            </Button>
          ) : null}
        </View>
      </ScrollView>
    </SafeAreaView>
  )
}
