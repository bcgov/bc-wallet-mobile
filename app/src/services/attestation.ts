import { AnonCredsCredentialOffer } from '@credo-ts/anoncreds'
import {
  Agent,
  BaseEvent,
  CredentialState,
  CredentialEventTypes,
  CredentialExchangeRecord,
  ProofState,
  ProofEventTypes,
  ProofExchangeRecord,
  BaseLogger,
  ConnectionRecord,
} from '@credo-ts/core'
import { BifoldError, BifoldAgent } from '@hyperledger/aries-bifold-core'
import {
  generateKey,
  appleAttestation,
  googleAttestation,
  isPlayIntegrityAvailable,
} from '@hyperledger/aries-react-native-attestation'
import { DeviceEventEmitter, Platform } from 'react-native'
import { getVersion, getBuildNumber, getSystemName, getSystemVersion } from 'react-native-device-info'
import { Subscription } from 'rxjs'

import { removeExistingInvitationIfRequired } from '../helpers/BCIDHelper'
import { credentialsMatchForProof } from '../helpers/credentials'
import { AttestationRequestParams, AttestationResult, requestNonceDrpc, requestAttestationDrpc } from '../helpers/drpc'

const defaultResponseTimeoutInMs = 10000 // DRPC response timeout

export type AttestationMonitorOptions = {
  attestationInviteUrl: string
  attestationCredDefIds: string[]
}

export const AttestationEventTypes = {
  Started: 'AttestationEvent.Started',
  Completed: 'AttestationEvent.Completed',
  FailedHandleOffer: 'AttestationEvent.FailedHandleOffer',
  FailedHandleProof: 'AttestationEvent.FailedHandleProof',
  FailedRequestCredential: 'AttestationEvent.FailedRequestCredential',
} as const

export interface AttestationCredentialFormat {
  attributes: {
    attestationInfo: []
  }
}

interface IndyRequest {
  indy: {
    requested_attributes?: {
      attestationInfo?: {
        names: string[]
        restrictions: { cred_def_id: string }[]
      }
    }
  }
}

interface AnonCredsRequest {
  anoncreds: {
    requested_attributes?: {
      attestationInfo?: {
        names: string[]
        restrictions: { cred_def_id: string }[]
      }
    }
  }
}

interface AttestationProofRequestFormat {
  request: IndyRequest & AnonCredsRequest
}

const AttestationErrorCodes = {
  BadInvitation: 2027,
  ReceiveInvitationError: 2028,
  GeneralProofError: 2029,
  FailedToConnectToAttestationAgent: 2030,
  FailedToFetchNonceForAttestation: 2031,
  FailedToGenerateAttestation: 2032,
  FailedToRequestAttestation: 2033,
  FailedToValidateAttestation: 2034,
} as const

export const isProofRequestingAttestation = async (
  proof: ProofExchangeRecord,
  agent: BifoldAgent,
  attestationCredDefIds: string[]
): Promise<boolean> => {
  const format = (await agent.proofs.getFormatData(proof.id)) as unknown as AttestationProofRequestFormat
  const formatToUse = format.request?.anoncreds ? 'anoncreds' : 'indy'

  return !!format.request?.[formatToUse]?.requested_attributes?.attestationInfo?.restrictions?.some((rstr) =>
    attestationCredDefIds ? attestationCredDefIds.includes(rstr.cred_def_id) : false
  )
}

export class AttestationMonitor {
  private proofSubscription?: Subscription
  private offerSubscription?: Subscription
  private agent: Agent
  private options: AttestationMonitorOptions
  private log?: BaseLogger
  private _attestationWorkflowInProgress = false

  // take in options, agent, and logger. Options should include the attestation service URL
  // and the proof to watch for along with the cred_ef_id of the attestation credentials.
  public constructor(agent: Agent, logger: BaseLogger, options: AttestationMonitorOptions) {
    this.agent = agent
    this.log = logger
    this.options = options
  }

  public get attestationWorkflowInProgress() {
    return this._attestationWorkflowInProgress
  }

  public async start(): Promise<void> {
    this.proofSubscription = this.agent.events
      .observable(ProofEventTypes.ProofStateChanged)
      .subscribe(this.handleProofStateChanged)

    this.offerSubscription = this.agent.events
      .observable(CredentialEventTypes.CredentialStateChanged)
      .subscribe(this.handleCredentialStateChanged)
  }

  public stop(): void {
    this.proofSubscription?.unsubscribe()
    this.offerSubscription?.unsubscribe()
  }

  public requestAttestationCredential = async () => {
    this.log?.info('Fetching attestation credential')

    this._attestationWorkflowInProgress = true
    DeviceEventEmitter.emit(AttestationEventTypes.Started)

    try {
      const connection = await this.connectToAttestationAgent()
      if (!connection) {
        throw new BifoldError(
          'Attestation Service',
          'Unable to connect to the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToConnectToAttestationAgent
        )
      }

      const nonce = await this.fetchNonceForAttestation(connection)
      if (!nonce) {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToFetchNonceForAttestation
        )
      }

      const attestationObj = await this.generateAttestation(nonce)
      if (!attestationObj) {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToGenerateAttestation
        )
      }

      const result = await this.requestAttestation(connection, attestationObj)
      if (result.status !== 'success') {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToValidateAttestation
        )
      }
    } catch (error) {
      this._attestationWorkflowInProgress = false
      this.log?.error('Failed to fetch attestation credential', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedRequestCredential, error)
    }
  }

  private handleCredentialStateChanged = async (event: BaseEvent) => {
    const { credentialRecord } = event.payload
    const credential = credentialRecord as CredentialExchangeRecord

    this.log?.info('Handling credential offer')

    try {
      const { offer } = await this.agent.credentials.getFormatData(credential.id)
      const offerData = (offer?.anoncreds ?? offer?.indy) as AnonCredsCredentialOffer

      // do nothing if not an attestation credential
      const { attestationCredDefIds } = this.options
      if (!attestationCredDefIds.includes(offerData?.cred_def_id ?? '')) {
        return
      }

      // if it's a new offer, automatically accept
      if (credential.state === CredentialState.OfferReceived) {
        this.log?.info('Accepting credential offer')
        await this.agent.credentials.acceptOffer({
          credentialRecordId: credential.id,
        })
      }

      // only finish loading state once credential is fully accepted
      if (credential.state === CredentialState.Done) {
        // TODO: credential.offer in flight completed
        this.log?.info('Credential accepted')

        this._attestationWorkflowInProgress = false

        DeviceEventEmitter.emit(AttestationEventTypes.Completed)
      }
    } catch (error) {
      this.log?.error('Failed to handle credential offer', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedHandleOffer, error)
    }
  }

  private handleProofStateChanged = async (event: BaseEvent) => {
    const { proofRecord } = event.payload
    const proof = proofRecord as ProofExchangeRecord

    this.log?.info('Handling proof received')

    if (proof.state !== ProofState.RequestReceived) {
      return
    }

    this.log?.info('Checking if proof is requesting attestation')

    try {
      // 1. Is the proof requesting an attestation credential
      if (!(await isProofRequestingAttestation(proof, this.agent, this.options.attestationCredDefIds))) {
        return
      }

      this.log?.info('Proof is requesting attestation')

      // 2. Does the wallet have a valid attestation credential that will
      // satisfy the proof request?
      const required = await this.attestationCredentialRequired(this.agent, proof.id)

      // 3. If yes, do nothing
      if (!required) {
        this.log?.info('Valid credentials already exist, nothing to do')

        return
      }

      // 4. If no, get a new attestation credential
      await this.requestAttestationCredential()
    } catch (error) {
      this.log?.error('Failed to handle proof', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedHandleProof, error)
    }
  }

  private async connectToAttestationAgent(): Promise<ConnectionRecord | undefined> {
    const invite = await this.agent.oob.parseInvitation(this.options.attestationInviteUrl)

    if (!invite) {
      this.log?.error('Unable to parse attestation agent invitation')

      throw new BifoldError(
        'Attestation Service',
        'Unable to parse the attestation agent invitation',
        'No details provided.',
        AttestationErrorCodes.BadInvitation
      )
    }

    this.log?.info('Removing existing invitation if required')
    await removeExistingInvitationIfRequired(this.agent, invite.id)

    this.log?.info('Receiving invitation')
    const { connectionRecord } = await this.agent.oob.receiveInvitation(invite)
    if (!connectionRecord) {
      throw new BifoldError(
        'Attestation Service',
        'Unable to accept attestation agent invitation',
        'No details provided.',
        AttestationErrorCodes.BadInvitation
      )
    }

    // this step will fail if there is more than one active connection record between a given wallet and
    // the traction instance which is why we need to `removeExistingInvitationIfRequired` above
    return await this.agent.connections.returnWhenIsConnected(connectionRecord.id)
  }

  private async fetchNonceForAttestation(connection: ConnectionRecord): Promise<string> {
    this.log?.info('Requesting nonce from controller')

    const requestNonceCb = await requestNonceDrpc(this.agent, connection)
    const nonceResponse = await requestNonceCb(defaultResponseTimeoutInMs)

    if (!nonceResponse) {
      this.log?.error('Failed to fetch nonce for attestation, code = none, reason = timeout occurred.')

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        'Timeout occurred.',
        AttestationErrorCodes.FailedToFetchNonceForAttestation
      )
    }

    this.log?.info('DRPC nonce response received')

    if (nonceResponse.error) {
      this.log?.error(
        `Failed to fetch nonce for attestation, code = ${nonceResponse.error.code}, reason = ${nonceResponse.error.message}`
      )

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        nonceResponse.error.message ?? 'No details provided.',
        AttestationErrorCodes.FailedToFetchNonceForAttestation
      )
    }

    const nonce = nonceResponse.result.nonce

    return nonce
  }

  private async requestAttestation(
    connection: ConnectionRecord,
    attestationObj: AttestationRequestParams
  ): Promise<AttestationResult> {
    this.log?.info('Requesting attestation credential from controller')

    const requestAttestationCb = await requestAttestationDrpc(this.agent, connection, attestationObj)
    const attestationResponse = await requestAttestationCb(defaultResponseTimeoutInMs)

    if (!attestationResponse) {
      this.log?.error('Failed to request attestation, code = none, reason = timeout occurred.')

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        'Timeout occurred.',
        AttestationErrorCodes.FailedToRequestAttestation
      )
    }

    this.log?.info('DRPC attestation response received')

    if (attestationResponse.error) {
      this.log?.error(
        `Failed to request attestation, code = ${attestationResponse.error.code}, reason = ${attestationResponse.error.message}`
      )

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        attestationResponse.error.message ?? 'No details provided.',
        AttestationErrorCodes.FailedToRequestAttestation
      )
    }

    return attestationResponse.result
  }

  private commonAttestationMessageComponent() {
    const common: Partial<AttestationRequestParams> = {
      app_version: `${getVersion()}-${getBuildNumber()}`,
      os_version: `${getSystemName()} ${getSystemVersion()}`,
    }

    return common
  }

  private async generateAttestation(nonce: string) {
    switch (Platform.OS) {
      case 'ios':
        return this.generateAppleAttestation(nonce)
      case 'android':
        return this.generateGoogleAttestation(nonce)

      default:
        // TODO(jl): throw unsupported platform error
        break
    }
  }

  private async generateAppleAttestation(nonce: string) {
    const common = this.commonAttestationMessageComponent()
    const shouldCacheKey = false

    this.log?.info('Generating key for Apple')
    const keyId = await generateKey(shouldCacheKey)

    this.log?.info('Using Apple on-device attestation')
    const attestationAsBuffer = await appleAttestation(keyId, nonce)
    const attestationRequest = {
      ...common,
      platform: 'apple',
      key_id: keyId,
      attestation_object: attestationAsBuffer.toString('base64'),
    } as AttestationRequestParams

    this.log?.info('On-device Apple attestation complete')

    return attestationRequest
  }

  private async generateGoogleAttestation(nonce: string) {
    const common = this.commonAttestationMessageComponent()

    this.log?.info('Checking if Play Integrity is available')

    const available = await isPlayIntegrityAvailable()
    if (!available) {
      return null
    }

    this.log?.info('Using Google on-device attestation')

    const tokenString = await googleAttestation(nonce)
    const attestationRequest = {
      ...common,
      platform: 'google',
      attestation_object: tokenString,
    } as AttestationRequestParams

    this.log?.info('On-device Google attestation complete')

    return attestationRequest
  }

  private attestationCredentialRequired = async (agent: BifoldAgent, proofId: string): Promise<boolean> => {
    agent.config.logger.info('Fetching proof by id')
    const proof = await agent?.proofs.getById(proofId)
    agent.config.logger.info('Second check if proof is requesting attestation')

    agent.config.logger.info('Checking if credentials match for proof request')
    const credentials = await credentialsMatchForProof(agent, proof)

    if (!credentials) {
      return true
    }

    // TODO:(jl) Should we be checking the length of the attributes matches some
    // expected length in the proof request?
    const format = (credentials.proofFormats.anoncreds ?? credentials.proofFormats.indy) as AttestationCredentialFormat
    if (format) {
      return format.attributes.attestationInfo.length === 0
    }

    return false
  }
}
