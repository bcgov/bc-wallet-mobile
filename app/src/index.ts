import {
  translationResources,
  ConfigurationContext,
  types,
  Record,
  indyLedgers,
  defaultConfiguration,
} from 'aries-bifold'
import merge from 'lodash.merge'
import { ReducerAction } from 'react'

import bundles from './assets/branding/credential-branding'
import AddCredentialButton from './components/AddCredentialButton'
import AddCredentialSlider from './components/AddCredentialSlider'
import EmptyList from './components/EmptyList'
import HomeContentView from './components/HomeContentView'
import { useNotifications } from './hooks/notifications'
import en from './localization/en'
import { proofRequestTemplates } from './request-templates'
import Developer from './screens/Developer'
import { pages } from './screens/OnboardingPages'
import PersonCredential from './screens/PersonCredential'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import { BCDispatchAction } from './store'
import { defaultTheme as theme } from './theme'

const localization = merge({}, translationResources, {
  en: { translation: en },
})

const selectedLedgers = indyLedgers.filter((item) => !item.id.startsWith('Indicio'))
const configuration: ConfigurationContext = {
  ...defaultConfiguration,
  pages,
  splash: Splash,
  terms: Terms,
  homeContentView: HomeContentView,
  credentialListHeaderRight: AddCredentialButton,
  credentialListOptions: AddCredentialSlider,
  credentialEmptyList: EmptyList,
  developer: Developer,
  OCABundleResolver: new types.oca.OCABundleResolver(bundles as unknown as types.oca.Bundles),
  record: Record,
  indyLedgers: selectedLedgers,
  settings: [],
  customNotification: {
    component: PersonCredential,
    onCloseAction: (dispatch?: React.Dispatch<ReducerAction<any>>) => {
      if (dispatch) {
        dispatch({
          type: BCDispatchAction.PERSON_CREDENTIAL_OFFER_DISMISSED,
          payload: [{ personCredentialOfferDismissed: true }],
        })
      }
    },
    pageTitle: 'PersonCredential.PageTitle',
    title: 'PersonCredentialNotification.Title',
    description: 'PersonCredentialNotification.Description',
    buttonTitle: 'PersonCredentialNotification.ButtonTitle',
  },
  useCustomNotifications: useNotifications,
  proofRequestTemplates,
  enableTours: true,
}

export default { theme, localization, configuration }
