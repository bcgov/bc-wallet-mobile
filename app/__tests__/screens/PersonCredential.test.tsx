import { StoreProvider } from '@hyperledger/aries-bifold-core'
import { render } from '@testing-library/react-native'
import React from 'react'

import PersonCredential from '../../src/screens/PersonCredential'
import { AttestationProvider } from '../../src/services/attestation'
import { initialState, reducer } from '../../src/store'

const mockNavigation = jest.fn()
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => {
    return mockNavigation
  },
}))

describe('Person Credential Screen', () => {
  beforeEach(() => {
    // Silence console.error because it will print a warning about Switch
    // "Warning: dispatchCommand was called with a ref ...".
    jest.spyOn(console, 'error').mockImplementation(() => {
      return null
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('screen renders correctly', () => {
    const tree = render(
      <StoreProvider initialState={initialState} reducer={reducer}>
        <AttestationProvider>
          <PersonCredential />
        </AttestationProvider>
      </StoreProvider>
    )

    expect(tree).toMatchSnapshot()
  })
})
