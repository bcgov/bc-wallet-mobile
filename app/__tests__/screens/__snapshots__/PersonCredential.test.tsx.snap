// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person Credential Screen screen renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
    ]
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
        "flexGrow": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "marginVertical": 20,
          }
        }
      >
        < />
      </View>
      <View
        style={
          {
            "backgroundColor": "#313132",
            "flexGrow": 1,
            "marginTop": 10,
            "paddingHorizontal": 25,
            "paddingVertical": 24,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "paddingBottom": 10,
            }
          }
        >
          <Text
            accessibilityRole="header"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 26,
                  "fontWeight": "bold",
                },
                {
                  "color": "#FFFFFF",
                  "flexShrink": 1,
                },
              ]
            }
          >
            PersonCredential.InstallServicesCard
          </Text>
        </View>
        <View
          style={
            {
              "marginTop": 10,
            }
          }
        >
          <View
            accessibilityLabel="PersonCredential.InstallApp"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#42803E",
                "borderRadius": 4,
                "opacity": 1,
                "padding": 16,
              }
            }
            testID="com.ariesbifold:id/InstallApp"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontSize": 18,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                    false,
                    false,
                    false,
                  ]
                }
              >
                PersonCredential.InstallApp
              </Text>
            </View>
          </View>
          <View
            accessibilityLabel="PersonCredential.AppOnOtherDevice"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "display": "flex",
                "marginTop": 10,
                "opacity": 1,
              }
            }
            testID="com.ariesbifold:id/AppOnOtherDevice"
          >
            <Text
              style={
                {
                  "color": "#42803E",
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              PersonCredential.AppOnOtherDevice
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "backgroundColor": "#313132",
            "flexGrow": 1,
            "marginTop": 10,
            "paddingHorizontal": 25,
            "paddingVertical": 24,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "paddingBottom": 10,
            }
          }
        >
          <Text
            accessibilityRole="header"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 26,
                  "fontWeight": "bold",
                },
                {
                  "color": "rgba(53, 130, 63, 0.35)",
                  "flexShrink": 1,
                },
              ]
            }
          >
            PersonCredential.CreatePersonCred
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
